flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/utils/PathUtils.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.file.table.utils;
+
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.util.UUID;
+
+/** Path utils for file system. */
+public class PathUtils {
+
+    public static Path getStagingPath(Path path) {
```

Comment:
```suggestion
    
public static Path getStagingPath(Path path) {
        return getStagingPath(
                path,
                () ->
                        String.format(
                                ".staging_%d_%s", System.currentTimeMillis(), UUID.randomUUID()));
    }

    @VisibleForTesting
    static Path getStagingPath(Path path, Supplier<String> suffixSupplier) {
```
The current test is "theoretically" flaky. We change that by using a callback and checking for the Precondition. That would enable us to create the following test:

```java
    @Test
    void testReusingStagingDirFails(@TempDir Path tmpDir) throws IOException {
        final String subfolderName = "directory-name";
        Files.createDirectory(tmpDir.resolve(subfolderName));
        assertThatThrownBy(
                        () ->
                                PathUtils.getStagingPath(
                                        org.apache.flink.core.fs.Path.fromLocalFile(
                                                tmpDir.toFile()),
                                        () -> subfolderName))
                .isEqualTo(IllegalStateException.class);
    }
```
And that would reveal that our assumption was actually wrong that `mkdirs` returns `false` if the directory already exists. :-/

Reply:
Hi @XComp, I'm afraid that this way also violates the arch unit test.

Here's what I tested in the local.

```java
public class PathUtils {

    public static Path getStagingPath(Path path) {
        // Add a random UUID to prevent multiple sinks from sharing the same staging dir.
        // Please see FLINK-29114 for more details
        return getStagingPath(
                path,
                () ->
                        String.format(
                                ".staging_%d_%s", System.currentTimeMillis(), UUID.randomUUID()));
    }

    @VisibleForTesting
    static Path getStagingPath(Path path, Supplier<String> suffixSupplier) {
        Path stagingDir = path.suffix(suffixSupplier.get());
        try {
            FileSystem fs = stagingDir.getFileSystem();
            Preconditions.checkState(
                    !fs.exists(stagingDir), "Staging dir %s already exists", stagingDir);
            fs.mkdirs(stagingDir);
            return stagingDir;
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}
```


Run `ArchitectureTest` to get the following stacktrace. (The first rule has been violated before, so let's leave that.)

```
java.lang.AssertionError: Architecture Violation [Priority: MEDIUM] - Rule 'Connector production code must depend only on public API when outside of connector packages' was violated (2 times):
Method <org.apache.flink.connector.file.table.utils.PathUtils.getStagingPath(org.apache.flink.core.fs.Path, java.util.function.Supplier)> calls method <org.apache.flink.util.Preconditions.checkState(boolean, java.lang.String, [Ljava.lang.Object;)> in (PathUtils.java:48)
Method <org.apache.flink.connector.file.table.utils.PathUtils.getStagingPath(org.apache.flink.core.fs.Path, java.util.function.Supplier)> is annotated with <org.apache.flink.annotation.VisibleForTesting> in (PathUtils.java:0)
```

Reply:
1. Generate duplicate code for temporary file creation in `FileSystemOutputFormat` and `HiveTableSink`. (similar to the base code version of this PR) which should be ok because it's not really complex code that we're dealing with, here.
2. Move the folder generation into a dedicated method that needs to be annotated as `@Public` to honor the ArchUnit tests. That would be a viable solution if we think that the temporary folder is in some means relevant to the user (e.g. is the naming scheme for the staging directory relevant so that it should be not changed between releases?). In that case, we should, indeed, annotate it as `@Public`.

To me, it sounds like the staging directory generation is an internal matter. How important is it that the naming scheme is matched between the `FileSystemTableSink` and the `HiveTableSink`? I would lean towards making the temp file generation internal functionality that is independent between `FileSystemTableSink` and `HiveTableSink` (option 1). But I'm not familiar with this code to judge here.

Reply:
I'm opting for option one as well. I didn't choose it in the first place because

<1> Not very sure if it is tolerable to have duplicate logic scattered throughout the codebase.

<2> I was hesitant about whether altering `FileSystemOutputFormat$Builder#setTempPath` to `FileSystemOutputFormat$Builder#setPath` might lead to any issues with dependencies. While I've manually reviewed the connector repositories and found no instances of the method being called, the impact on custom connectors remains unclear. Perhaps my concerns are a bit excessive.

It appears that we're on the same page. I'll address the comment as we've discussed.